File: src/swift/4_BehavioralPatterns/4_Iterator/Exercise.swift
import Foundation
import XCTest

class Node<T>
{
  let value: T
  var left: Node<T>? = nil
  var right: Node<T>? = nil
  var parent: Node<T>? = nil

  init(_ value: T)
  {
    self.value = value
  }

  init(_ value: T, _ left: Node<T>, _ right: Node<T>)
  {
    self.value = value
    self.left = left
    self.right = right

    left.parent = self
    right.parent = self
  }

  private func traverse(_ current: Node<T>, _ buffer: inout [T])
  {
    // not the optimal solution, but it works
    buffer.append(current.value)

    if let currentLeft = current.left
    {
      traverse(currentLeft, &buffer)
    }

    if let currentRight = current.right
    {
      traverse(currentRight, &buffer)
    }
  }

  public var preOrder: [T]
  {
    get {
      var buffer = [T]()
      traverse(self, &buffer)
      return buffer
    }
  }
}

class UMBaseTestCase : XCTestCase {}

class Evaluate: UMBaseTestCase
{
  func simpleTest()
  {
    let node = Node("a",
      Node("b",
        Node("c"),
        Node("d")
      ),
      Node("e")
    )
    XCTAssertEqual("abcde", node.preOrder.joined(),
      "Preorder traversal of this structure should yield 'abcde'")
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("simpleTest", simpleTest)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/python/4_BehavioralPatterns/4_Iterator/exercise.py
from unittest import *

class Node:
  def __init__(self, value, left=None, right=None):
    self.right = right
    self.left = left
    self.value = value

    self.parent = None

    if left:
      self.left.parent = self
    if right:
      self.right.parent = self

  def traverse_preorder(self):
    yield self.value
    if self.left:
      yield from self.left.traverse_preorder()
    if self.right:
      yield from self.right.traverse_preorder()
  # def traverse_preorder(self):
  #   def traverse(current):
  #     yield current
  #     if current.left:
  #       for left in traverse(current.left):
  #         yield left
  #     if current.right:
  #       for right in traverse(current.right):
  #         yield right
  #   for node in traverse(self):
  #     yield node.value

class Evaluate(TestCase):
  def test_exercise(self):
    node = Node('a',
                Node('b',
                     Node('c'),
                     Node('d')),
                Node('e'))
    self.assertEqual(
      'abcde',
      ''.join([x for x in node.traverse_preorder()])
    )